### ISSUE TITLE
Circular import with allennlp.modules

### ISSUE THREAD

**Issue by nelson-liu** [allenai/allennlp#2250]
**System (please complete the following information):**
 - OS: OSX
 - Python version: 3.6.6
 - AllenNLP version: master
 - PyTorch version: 0.4.1

**Question**
Maybe I'm missing something, but importing from modules seems to break on my machine, but only from the REPL. Seems like tests work fine?

```
$ ipython
Python 3.6.6 |Anaconda, Inc.| (default, Jun 28 2018, 11:07:29)
Type 'copyright', 'credits' or 'license' for more information
IPython 6.5.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import torch
   ...: from torch.autograd import Variable
   ...: from allennlp.modules import Pruner
   ...:
   ...:
Deprecated cache directory found (/Users/nfliu/.allennlp/datasets).  Please remove this directory from your system to free up space.
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-1-340a85315d8a> in <module>()
      1 import torch
      2 from torch.autograd import Variable
----> 3 from allennlp.modules import Pruner

~/Documents/Github/allennlp/allennlp/modules/__init__.py in <module>()
     17 from allennlp.modules.similarity_functions import SimilarityFunction
     18 from allennlp.modules.pruner import Pruner
---> 19 from allennlp.modules.text_field_embedders import TextFieldEmbedder
     20 from allennlp.modules.time_distributed import TimeDistributed
     21 from allennlp.modules.token_embedders import TokenEmbedder, Embedding

~/Documents/Github/allennlp/allennlp/modules/text_field_embedders/__init__.py in <module>()
      7
      8 from allennlp.modules.text_field_embedders.text_field_embedder import TextFieldEmbedder
----> 9 from allennlp.modules.text_field_embedders.basic_text_field_embedder import BasicTextFieldEmbedder

~/Documents/Github/allennlp/allennlp/modules/text_field_embedders/basic_text_field_embedder.py in <module>()
     10 from allennlp.modules.text_field_embedders.text_field_embedder import TextFieldEmbedder
     11 from allennlp.modules.time_distributed import TimeDistributed
---> 12 from allennlp.modules.token_embedders.token_embedder import TokenEmbedder
     13
     14

~/Documents/Github/allennlp/allennlp/modules/token_embedders/__init__.py in <module>()
     10 from allennlp.modules.token_embedders.openai_transformer_embedder import OpenaiTransformerEmbedder
     11 from allennlp.modules.token_embedders.bert_token_embedder import BertEmbedder, PretrainedBertEmbedder
---> 12 from allennlp.modules.token_embedders.bidirectional_language_model_token_embedder import \
     13         BidirectionalLanguageModelTokenEmbedder

~/Documents/Github/allennlp/allennlp/modules/token_embedders/bidirectional_language_model_token_embedder.py in <module>()
      6 from allennlp.common.checks import ConfigurationError
      7 from allennlp.data import TokenIndexer, Token
----> 8 from allennlp.models.archival import load_archive
      9 from allennlp.modules.scalar_mix import ScalarMix
     10 from allennlp.modules.token_embedders.token_embedder import TokenEmbedder

~/Documents/Github/allennlp/allennlp/models/__init__.py in <module>()
      6 from allennlp.models.model import Model
      7 from allennlp.models.archival import archive_model, load_archive, Archive
----> 8 from allennlp.models.biattentive_classification_network import BiattentiveClassificationNetwork
      9 from allennlp.models.constituency_parser import SpanConstituencyParser
     10 from allennlp.models.biaffine_dependency_parser import BiaffineDependencyParser

~/Documents/Github/allennlp/allennlp/models/biattentive_classification_network.py in <module>()
     10 from allennlp.common.checks import check_dimensions_match, ConfigurationError
     11 from allennlp.data import Vocabulary
---> 12 from allennlp.modules import Elmo, FeedForward, Maxout, Seq2SeqEncoder, TextFieldEmbedder
     13 from allennlp.models.model import Model
     14 from allennlp.nn import InitializerApplicator, RegularizerApplicator

ImportError: cannot import name 'TextFieldEmbedder'
```

**Comment by nelson-liu** [comment]
ah, i see that @brendan-ai2 already put something to make it work for typechecking ( https://github.com/allenai/allennlp/blob/master/allennlp/modules/token_embedders/bidirectional_language_model_token_embedder.py#L13-L15 ). I think this needs to be a bit broader, though, as it currently breaks for anyone using allennlp as a library.

**Comment by brendan-ai2** [comment]
Fix in https://github.com/allenai/allennlp/pull/2257. I don't understand why this was only an issue when using `ipython`. If any Python gurus would like to chime in, I'd appreciate it!

**Comment by nelson-liu** [comment]
Thanks for the fix @brendan-ai2 , we should probably make a new release after the holidays with this fix?

Maybe this was only an issue when using `python` because when using the `allennlp` commands the importing occurs in a different order that doesn't result in a circular dependency (since `load_archive` or something is already imported earlier)?

**Comment by brendan-ai2** [comment]
Yes, after the holidays is probably best.

Ah, that makes sense. Thanks! So when you import a model first, you also import `load_archive` via allennlp/allennlp/models/__init__.py. I checked this by running the following on master before #2257:

```
$ ipython
In [1]: import torch
In [2]: from torch.autograd import Variable
In [3]: from allennlp.models import Event2Mind
In [4]: from allennlp.modules import Pruner
In [5]:
```

We should find a way to test for or systematically prevent this sort of import brittleness.

**Comment by matt-gardner** [comment]
Fixed by #2257.  (re-open if I missed some reason that you didn't close this earlier)
